### ISSUE TITLE
Migration action throws an exception that no callbacks are registered

### ISSUE THREAD

**Issue by engahmeds3ed** [Prospress/action-scheduler#948]
We face the following error:

> action failed via Async Request: Scheduled action for action_scheduler/migration_hook will not be executed as no callbacks are registered.

![image](https://user-images.githubusercontent.com/15707971/236234650-131abaea-c8c2-4f4c-8e07-cf1185e21119.png)

and in the same time we see the following error in log:

```
[03-May-2023 20:09:02 UTC] PHP Fatal error:  Uncaught RuntimeException: Error saving action: Error saving action: Database error. in /www/XXX/public/wp-content/plugins/XPLUGIN/inc/Dependencies/ActionScheduler/classes/migration/ActionScheduler_DBStoreMigrator.php:44
Stack trace:
#0 /www/XXX/public/wp-content/plugins/XPLUGIN/inc/Dependencies/ActionScheduler/classes/data-stores/ActionScheduler_HybridStore.php(242): ActionScheduler_DBStoreMigrator->save_action(Object(ActionScheduler_Action), NULL)
#1 /www/XXX/public/wp-content/plugins/XPLUGIN/inc/Dependencies/ActionScheduler/classes/ActionScheduler_ActionFactory.php(244): ActionScheduler_HybridStore->save_action(Object(ActionScheduler_Action))
#2 /www/XXX/public/wp-content/plugins/XPLUGIN/inc/Dependencies/ActionScheduler/classes/ActionScheduler_ActionFactory.php(114): ActionScheduler_ActionFactory->store(Object(ActionScheduler_Action))
#3 /www/XXX/public/wp-content/plugins/XPLUGIN/inc/Dependencies/ActionScheduler/functions.php(81): ActionScheduler_ActionFactory->single_unique('action_schedule...', Array, 1683144602, 'action-schedule...', false)
#4 /www/XXX/public/wp-content/plugins/XPLUGIN/inc/Dependencies/ActionScheduler/classes/migration/Scheduler.php(89): as_schedule_single_action(1683144602, 'action_schedule...', Array, 'action-schedule...')
#5 /www/XXX/public/wp-content/plugins/XPLUGIN/inc/Dependencies/ActionScheduler/classes/migration/Controller.php(119): Action_Scheduler\Migration\Scheduler->schedule_migration()
#6 /www/XXX/public/wp-includes/class-wp-hook.php(308): Action_Scheduler\Migration\Controller->schedule_migration('')
#7 /www/XXX/public/wp-includes/class-wp-hook.php(332): WP_Hook->apply_filters(NULL, Array)
#8 /www/XXX/public/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#9 /www/XXX/public/wp-settings.php(645): do_action('wp_loaded')
#10 /www/XXX/public/wp-config.php(101): require_once('/www/...')
#11 /www/XXX/public/wp-load.php(50): require_once('/www/...')
#12 /www/XXX/public/wp-admin/admin-ajax.php(22): require_once('/www/...')
#13 {main}
  thrown in /www/XXX/public/wp-content/plugins/XPLUGIN/inc/Dependencies/ActionScheduler/classes/migration/ActionScheduler_DBStoreMigrator.php on line 44
```

We are loading the action scheduler code by including the main file directly in our plugin's code without wrapping it in any action.

I don't know if it's related to our plugin or any other plugin that uses Action Scheduler.

Is this a known issue or I'm doing something wrong from my side?

Thanks.

**Comment by barryhughes** [comment]
Hi @engahmeds3ed! I can't immediately replicate, even if I test with all other plugins deactivated except for Imagify  (the free version distributed in the plugin directory, that is).

> I don't know if it's related to our plugin or any other plugin that uses Action Scheduler.

To clarify: are you testing while other plugins are active, and have you been able to replicate on your side when *only* Imagify is active?

**Comment by engahmeds3ed** [comment]
Thanks @barryhughes for checking this one.

No this is a weird behavior in a client's website and I was testing while at least there were 3 active plugins that use Action Scheduler.

I added my question here to try to know more from you what may cause this issue, but I didn't see it in any other site.

**Comment by barryhughes** [comment]
Hmm, tricky, especially if neither of us can reproduce 🤔 

> I was testing while at least there were 3 active plugins that use Action Scheduler.

The general idea is that only one instance of the library will actually be loaded, and it should be the highest available version. However, if one of the plugins includes Action Scheduler 3.2.0 or earlier, it's possible that an older version is inadvertently being loaded instead [(see this developer advisory)](https://developer.woocommerce.com/2021/10/12/best-practices-for-deconflicting-different-versions-of-action-scheduler/).

That could lead to a range of problems, from the re-emergence of bugs we subsequently solved or possibly other issues that might result from a schema mismatch. I'm not sure what level of access you have to the client's website, but is that a possibility (does one of those plugins include 3.2.0 or earlier)?

**Comment by engahmeds3ed** [comment]
After deep investigation on that, I found that the problem is coming from this sql query
```
INSERT INTO wp_actionscheduler_actions ( `hook`, `status`, `scheduled_date_gmt`, `scheduled_date_local`, `schedule`, `group_id`, `priority`, `args` )
SELECT 'action_scheduler/migration_hook', 'pending', '2023-05-15 10:51:09', '2023-05-15 03:51:09', 'O:30:\\"ActionScheduler_SimpleSchedule\\":2:{s:22:\\"\\0*\\0scheduled_timestamp\\";i:1684147869;s:41:\\"\\0ActionScheduler_SimpleSchedule\\0timestamp\\";i:1684147869;}', 1, 10, '[]' FROM DUAL
WHERE ( SELECT NULL FROM DUAL ) IS NULL
```

which is coming from here:

https://github.com/woocommerce/action-scheduler/blob/4ea5ba6ba57f9589a4222472e2baac85ab2b2c53/classes/data-stores/ActionScheduler_DBStore.php#L148-L156

# STILL CHECKING...

**Server details:**


-------------------------
Server version | 10.5.17-MariaDB-1:10.5.17+maria~ubu2004
-- | --
Client version | mysqlnd 8.0.22

What do u think about that?

**Comment by barryhughes** [comment]
Interesting. Did you capture a specific error message along with the problematic query—on first look, it seems like it should be valid?

Also, to confirm, does this remain something that you cannot reproduce outside of this one client website?

**Comment by engahmeds3ed** [comment]
Yes it's a valid query.

I added some logs in this line: 

https://github.com/woocommerce/action-scheduler/blob/4ea5ba6ba57f9589a4222472e2baac85ab2b2c53/classes/data-stores/ActionScheduler_DBStore.php#L118

Plz note that now I can reproduce the error in this specific server using the action scheduler standalone plugin with the first time activating the plugin so to see it I deactivate and activate the standalone plugin.

Those are the logs:

```
[15-May-2023 10:50:09 UTC] save_action_to_db_2: array (
  'action_id' => 0,
  'sql' => '
INSERT INTO wp_actionscheduler_actions ( `hook`, `status`, `scheduled_date_gmt`, `scheduled_date_local`, `schedule`, `group_id`, `priority`, `args` )
SELECT \'action_scheduler/migration_hook\', \'pending\', \'2023-05-15 10:51:09\', \'2023-05-15 03:51:09\', \'O:30:\\"ActionScheduler_SimpleSchedule\\":2:{s:22:\\"\\0*\\0scheduled_timestamp\\";i:1684147869;s:41:\\"\\0ActionScheduler_SimpleSchedule\\0timestamp\\";i:1684147869;}\', 1, 10, \'[]\' FROM DUAL
WHERE ( SELECT NULL FROM DUAL ) IS NULL',
)
[15-May-2023 10:50:09 UTC] PHP Fatal error:  Uncaught RuntimeException: Error saving action: Error saving action: Database error. in /www/XSITE/public/wp-content/plugins/action-scheduler/classes/migration/ActionScheduler_DBStoreMigrator.php:44
Stack trace:
#0 /www/XSITE/public/wp-content/plugins/action-scheduler/classes/data-stores/ActionScheduler_HybridStore.php(242): ActionScheduler_DBStoreMigrator->save_action(Object(ActionScheduler_Action), NULL)
#1 /www/XSITE/public/wp-content/plugins/action-scheduler/classes/ActionScheduler_ActionFactory.php(327): ActionScheduler_HybridStore->save_action(Object(ActionScheduler_Action))
#2 /www/XSITE/public/wp-content/plugins/action-scheduler/classes/ActionScheduler_ActionFactory.php(315): ActionScheduler_ActionFactory->store(Object(ActionScheduler_Action))
#3 /www/XSITE/public/wp-content/plugins/action-scheduler/functions.php(102): ActionScheduler_ActionFactory->create(Array)
#4 /www/XSITE/public/wp-content/plugins/action-scheduler/classes/migration/Scheduler.php(89): as_schedule_single_action(1684147869, 'action_schedule...', Array, 'action-schedule...')
#5 /www/XSITE/public/wp-content/plugins/action-scheduler/classes/migration/Controller.php(124): Action_Scheduler\Migration\Scheduler->schedule_migration()
#6 /www/XSITE/public/wp-includes/class-wp-hook.php(308): Action_Scheduler\Migration\Controller->schedule_migration('')
#7 /www/XSITE/public/wp-includes/class-wp-hook.php(332): WP_Hook->apply_filters(NULL, Array)
#8 /www/XSITE/public/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#9 /www/XSITE/public/wp-settings.php(645): do_action('wp_loaded')
#10 /www/XSITE/public/wp-config.php(101): require_once('/www/XSITE...')
#11 /www/XSITE/public/wp-load.php(50): require_once('/www/XSITE...')
#12 /www/XSITE/public/wp-admin/admin.php(34): require_once('/www/XSITE...')
#13 /www/XSITE/public/wp-admin/plugins.php(10): require_once('/www/XSITE...')
#14 {main}
  thrown in /www/XSITE/public/wp-content/plugins/action-scheduler/classes/migration/ActionScheduler_DBStoreMigrator.php on line 44
```

when running this query in the mysql server directly using adminer.php tool, it's working properly but when using a plugin that uses wpdb it fails and to make the query works I change the select single quotes to double quotes to be like

```
INSERT INTO wp_actionscheduler_actions ( `hook`, `status`, `scheduled_date_gmt`, `scheduled_date_local`, `schedule`, `group_id`, `priority`, `args` )
SELECT "action_scheduler/migration_hook", "pending", "2023-05-15 10:51:09", "2023-05-15 03:51:09", "O:30:\\\"ActionScheduler_SimpleSchedule\\\":2:{s:22:\\\"\\0*\\0scheduled_timestamp\\\";i:1684147869;s:41:\\\"\\0ActionScheduler_SimpleSchedule\\0timestamp\\\";i:1684147869;}", 1, 10, "[]" FROM DUAL
WHERE ( SELECT NULL FROM DUAL ) IS NULL
```

**Comment by barryhughes** [comment]
@engahmeds3ed using PhpMyAdmin or similar, can you run (updating the table prefix as needed):

```sql
SHOW COLUMNS IN wp_actionscheduler_actions
```

What is the result?

**Comment by engahmeds3ed** [comment]
@barryhughes 

![image](https://github.com/woocommerce/action-scheduler/assets/15707971/2d605eb9-53eb-404e-a5c2-c45040c7ac34)

**Comment by barryhughes** [comment]
It's a little puzzling. 

Presumably there is no database error (because of [this line](https://github.com/woocommerce/action-scheduler/blob/4ea5ba6ba57f9589a4222472e2baac85ab2b2c53/classes/data-stores/ActionScheduler_DBStore.php#L118): the exception only reads as *'Database error'* if WPDB didn't capture a more specific error) and, based on your logging—specifically `'action_id' => 0`—it seems the insert is failing.

> The query works I change the select single quotes to double quotes

That's also odd, since single quotes are the SQL standard (though MySQL [supports both conventions](https://dev.mysql.com/doc/refman/8.0/en/string-literals.html) for string literals). I'll try to make some time to replicate with MariaDB 10.5.17. Thanks for your patience in the meantime.

**Comment by barryhughes** [comment]
I'm afraid I still cannot replicate this. Here using a completely clean WP installation:

- Using WordPress 6.2.2 (via the official Docker image).
- Using MariaDB 10.5.17 (also via the official Docker image).
- I install and activate Query Monitor then Action Scheduler, no errors and no fatals/uncaught exceptions.
- I deactivate and reactivate Action Scheduler, and again see no problem.

Is there a step I'm missing and am I correct in thinking you so far can also only replicate in this one very specific environment (perhaps there is some aspect of the configuration we would need to duplicate)?

**Comment by barryhughes** [comment]
I'll go ahead and close the issue, since:

- We cannot replicate.
- It sounds like it the issue only surfaces on one specific website (suggesting a conflict, or else highly specific environmental conditions that we've been unable to identify).

If the problem occurs again, and especially if we can identify steps to replicate, we can definitely take another look.

**Comment by crstauf** [comment]
I'm getting this as well (on only one site that I'm aware of at the moment):

```
2024-12-18 00:18:12 +0000
action failed via WP CLI: Scheduled action for action_scheduler/migration_hook will not be executed as no callbacks are registered.
```

This results in the action being re-scheduled, failing again, and being re-scheduled, again.

**Comment by dwainm** [comment]
We should only re-open this issue again we have much better replication steps with the right context. We can't do anything without the ability to reproduce it.
